torch/distributed/fsdp/_flat_param.py

```diff
@@ -490,19 +498,39 @@ def __init__(
         self._use_full_prec_in_eval = (
             os.environ.get(_FSDP_USE_FULL_PREC_IN_EVAL, "") == "1"
         )
+        self._use_fake_all_gather = os.environ.get(_FSDP_USE_FAKE_ALL_GATHER, "") == "1"
+        self._use_fake_reduce = os.environ.get(_FSDP_USE_FAKE_REDUCE, "") == "1"
         if self._skip_writeback_check:
             _warn_skip_writeback_check(
                 log,
                 f"Since {_FSDP_SKIP_WRITEBACK_CHECK}=1, FSDP will not check "
                 "for parameter or gradient writeback. Changing parameter or "
                 "gradient storages may lead to silent correctness errors.",
             )
+        if self._use_fake_all_gather:
+            _warn_use_fake_all_gather(
+                log,
+                f"Since {_FSDP_USE_FAKE_ALL_GATHER}=1, FSDP will not execute "
+                "all-gather ops. Your training will be incorrect, but "
+                "can reveal how much time spent on all-gather ops.",
+            )
+        if self._use_fake_reduce:
+            _warn_use_fake_reduce(
+                log,
+                f"Since {_FSDP_USE_FAKE_REDUCE}=1, FSDP will not execute "
+                "reduce-scatter ops. Your training will be incorrect, but "
+                "can reveal how much time spent on reduce-scatter ops.",
+            )
         # Only align addresses for `use_orig_params=True` (for now)
         align_addresses = use_orig_params
         self._init_get_unflat_views_fn(align_addresses)
         self.device = device
         self._device_handle = _FSDPDeviceHandle.from_device(self.device)
         self.process_group = process_group
+        if self._use_fake_all_gather or self._use_fake_reduce:
+            self._fake_process_group = FakeProcessGroup(
```

Comment:
cc: @wanchaol Is there any concern of constructing 1 `FakeProcessGroup` for every FSDP instance (instead of reusing a single one)? It seems okay?

Reply:
It seems ok to reconstruct FakePG if needed. I am mainly wondering why we choose to create a FakePG inside FSDP, rather than create a FakePG, and pass it to FSDP? the latter approach don't need any FSDP specific changes IIUC

Reply:
Good point. @ckluk2 Can you try doing that?

(pass a `FakeProcessGroup` directly to FSDP's arg)
https://github.com/pytorch/pytorch/blob/9062e429db5bcba306a7ef5e1d6582bef564bca2/torch/distributed/fsdp/fully_sharded_data_parallel.py#L413

Reply:
* I actually prefer adding this feature to FSDP, because users would like to use this feature without changing this code
* Also, we will need to do some FSDP-specific change anyway, because we want to separately measure the performance impact of all_gather and reduce_scatter.